src/expr/src/relation/canonicalize.rs

```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
             predicates.push(expr);
         }
     }
+
+    // 3) Reduce across `predicates`.
+    // If a predicate `p` cannot be null, and `f(p)` is of type bool
+    // (`f(p)` is allowed to be nullable), then the predicate `p & f(p)` is
+    // equal to `p & f(true)`, and `!p & f(p)` is equal to `!p & f(false)`.
+    // For any index i, the `Vec` of predicates `[p1, ... pi, ... pn]` is
+    // equivalent to the single predicate `pi & (p1 & ... & p(i-1) & p(i+1)
+    // ... & pn)`.
+    // Thus, if `pi` (resp. `!pi`) cannot be null, it is valid to replace with
+    // `true` (resp. `false`) every subexpression in
+    // `(p1 & ... & p(i-1) & p(i+1) ... & pn)` that is equal to `pi`.
+
+    // (As for the reduction does not work if `p` can be null: if `p` is null,
+    // then p & f(p) becomes `null & f(null)`. `null & f(null)` is not
+    // necessarily equal to `null & f(true)` because `null & true = null`
+    // and `null & false = false`.)
+
+    // Note that this does some dedupping of predicates since if `p1 = p2`
+    // then this reduction process will replace `p1` with true.
+
+    // Pop out a predicate `p` from `predicates_to_apply`. Check all the other
+    // predicates (this is the union of `predicates_to_apply` and
+    // `applied_predicates`), and if `p` is a subexpression of any of those
+    // predicates, replace the subexpression with `true`. Then stick `p` into
+    // `applied_predicates`. If a predicate `q` in `applied_predicates` was
+    // altered in the aforementioned process, remove `q` from
+    // `applied_predicates` and put it in `predicates_to_apply` so we can reduce
+    // the predicates even further.
+    let mut predicates_to_apply = Vec::new();
+    let mut applied_predicates = Vec::<MirScalarExpr>::new();
+    // Seed `predicates_to_apply` with the `Vec` of all predicates.
+    std::mem::swap(&mut predicates_to_apply, predicates);
+
+    while let Some(predicate_to_apply) = predicates_to_apply.pop() {
+        let mut replace_subexpr_other_predicates =
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
```

Comment:
I think I understand this now, but the structure seems pretty convoluted. This is (correct me if wrong) the "worklist" pattern where things in the worklist are todo, and in the course of working things are dequeued, processed, and may involve putting things back in the queue. If that's right, I think this description (or something akin to it) is more helpful than the mechanical description up above (i.e. "take this out, put this in").

Also, it's pretty hard to grok with the closures that capture unclear state and .. idk whose names aren't quite enough to understand what they do. Like, I had to go read the implementation to determine that `visit_mut_replace_subexpr`'s main feature is that it doesn't process the `cond` block in `If` expressions.

Anyhow, I think this makes sense to me now, but I think each reader who comes to this will need to read all of the code to understand clearly what is happening. If you can add some comments at choice locations that reduces that property I'd be delighted!